### PR TITLE
linux: synthesize events for always-ready fds

### PR DESCRIPTION
**Pull request by vtjnash**
epoll rejects registration of file descriptors that are always ready for
I/O -- regular files, /dev/null, and similar character devices -- with
EPERM. libuv's io_uring-based epoll_ctl path aborted on that error, so
wrapping such an fd in a uv_pipe_t and reading from or writing to it
crashed the event loop. (uv_guess_handle() reports these as UV_FILE.)

Instead of aborting, park the watcher back on the loop's watcher_queue
and, on every iteration of uv__io_poll(), pre-populate the events[]
array with synthesized events for those descriptors before calling
epoll_pwait() (shrinking its range accordingly) so the normal dispatch
path delivers them. The watchers stay queued, so the next tick retries
the (failing) registration; uv__io_stop() evicts them via
uv__queue_remove() as usual.

The synthesized batch is limited to half of events[] so epoll_pwait() is
never called with maxevents == 0 and ordinary descriptors are not
starved, and the serviced batch is rotated to the tail of the queue so
that, when more descriptors are always-ready than fit at once, later
iterations pick up the ones skipped -- mirroring epoll's own
level-triggered round-robin.

Add test-pipe-dev-null.c covering both reading from and writing to the
null device (the write also exercises a trailing zero-length buffer). It
is skipped on Windows, where uv_pipe_open() does not accept the NUL
device.

Fix for #5182, although technically that is a user error so this is just a robustness change (failure to call uv_guess_handle will lead to UB, generally much worse on Windows than elsewhere).

### CODE REVIEW

**Comment by saghul** [comment]
How would users know they are "using it wrong" if we paper over the bad usage?

Crashing signals that the program is incorrect, they can fix it and move on.

I don't think we are helping here.

My 2 cents!

**Comment by vtjnash** [comment]
Does it matter?

**Comment by vtjnash** [comment]
(To be clear, I'm genuinely asking, since Linux epoll is the only unix backend where this doesn't work. Which is a pretty awkward since epoll promises to work the same as poll, but poll does handle this.)

**Comment by saghul** [comment]
Sure. My thinking is: even if we make it "work", it always be ready and this busy loop all the time.

Then users may ask themselves what's wrong, but because we covered the error it will be much harder to debug IMHO.

Since libuv provides enough APIs to "do it right", that abort signals your program is wrong and they can fix it properly.

**Comment by vtjnash** [comment]
Counterpoint: it isn't a busy loop: it is actually ready all the time, so it will finish, and remove itself from the loop immediately. All the other unix and even linux poll got this right (the unix philosophy: all pipes are just fd's and all fd's that represent read/write IO are implemented to work). Only linux epoll implemented this wrong (it documents that it is guaranteed to work on all fd that implement poll). Which also means only linux epoll (out of the unix) will abort here--every other unix will do the expected behavior and already works correctly.

**Comment by saghul** [comment]
What was the case on Windows?

**Comment by vtjnash** [comment]
It isn't a NamedPipe, so our explicit check for NamedPipe errors

File: src/unix/linux.c [hunk 1]
```diff
@@ -1369,6 +1383,30 @@ static void uv__epoll_ctl_flush(int epollfd,
...
+
+  if (fd < 0 || (unsigned) fd >= loop->nwatchers)
+    return;
+
+  w = loop->watchers[fd];
+  if (w == NULL)
+    return;
+
+  w->events = 0;
+  if (uv__queue_empty(&w->watcher_queue))
```
Comment (bnoordhuis) [inline src/unix/linux.c:1405]: Should be an assert? If the queue is not empty, then something is wrong, isn't it?

For that matter, the previous two checks cannot ever fail either, right?

File: src/unix/linux.c [hunk 2]
```diff
@@ -1476,23 +1534,70 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
...
+    if (r == -1)
       assert(errno == EINTR);
-    else if (nfds == 0)
+    else if (r == 0 && nready == 0)
       /* Unlimited timeout should only return with events or signal. */
       assert(timeout != -1);
 
-    if (nfds == 0 || nfds == -1) {
+    /* Fold the synthesized events in front of what epoll_pwait() returned. */
+    nfds = (r < 0 ? 0 : r) + nready;
```
Comment (bnoordhuis) [inline src/unix/linux.c:1591]:
```suggestion
    nfds = nready;
    if (r != -1)
      nfds += r;
```
Easier to read, IMO.

File: src/unix/linux.c [hunk 3]
```diff
@@ -1476,23 +1534,70 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
...
+      synthesized = 1;
+      /* Take a batch off the head of the queue, but reserve half of events[]
+       * for epoll_pwait() -- so it always gets a healthy number of slots (and
+       * never maxevents == 0) and real descriptors aren't starved. Rotate the
+       * batch to the tail so that, when more descriptors are always-ready than
+       * fit here, later ticks service the ones skipped now (epoll is
+       * level-triggered and round-robins its own overflow the same way).
+       */
+      while ((unsigned) nready < ARRAY_SIZE(events) / 2 &&
+             !uv__queue_empty(&loop->watcher_queue)) {
```
Comment (bnoordhuis) [inline src/unix/linux.c:1558]: The (by far) common case is that the queue is empty, right? Then that check should come first, otherwise it's always doing an unnecessary `x < y` check.

File: src/unix/linux.c [hunk 4]
```diff
@@ -1476,23 +1534,70 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
...
+        events[nready].events = w->pevents;
+        events[nready].data.fd = w->fd;
+        nready++;
+        uv__queue_insert_tail(&wq, q);
+      }
+      while (!uv__queue_empty(&wq)) {
+        q = uv__queue_head(&wq);
+        uv__queue_remove(q);
+        uv__queue_insert_tail(&loop->watcher_queue, q);
+      }
```
Comment (bnoordhuis) [inline src/unix/linux.c:1571]:
```suggestion
      uv__queue_add(&loop->watcher_queue, &wq);
```
O(1) vs O(n)
